source/SkiaSharp.Views/SkiaSharp.Views.WPF/SKGLElement.cs

```diff
@@ -0,0 +1,221 @@
+﻿using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using System.Windows;
+using System.Windows.Controls;
+using System.Windows.Media;
+using System.Windows.Media.Imaging;
+using OpenTK.Graphics;
+using System.Windows.Media.Media3D;
+using SkiaSharp.Views.Desktop;
+using OpenTK.Wpf;
+using SkiaSharp;
+using OpenTK.Graphics.OpenGL;
+using OpenTK.Platform.Windows;
+using OpenTK;
+using System.Windows.Interop;
+using OpenTK.Platform;
+#if NETCOREAPP || NET
+using OpenTK.Mathematics;
+#endif
+
+namespace SkiaSharp.Views.WPF
+{
+	[DefaultEvent("PaintSurface")]
+	[DefaultProperty("Name")]
+	public class SKGLElement : GLWpfControl, IDisposable
+	{
+		private const SKColorType colorType = SKColorType.Rgba8888;
+		private const GRSurfaceOrigin surfaceOrigin = GRSurfaceOrigin.BottomLeft;
+
+		private bool designMode;
+
+		private GRContext grContext;
+		private GRGlFramebufferInfo glInfo;
+		private GRBackendRenderTarget renderTarget;
+		private SKSurface surface;
+		private SKCanvas canvas;
+
+		private SKSizeI lastSize;
+
+		public SKGLElement()
+			: base()
+		{
+			Initialize();
+		}
+
+		private void Initialize()
+		{
+			designMode = DesignerProperties.GetIsInDesignMode(this);
+			var settings = new GLWpfControlSettings() { MajorVersion = 2, MinorVersion = 1, RenderContinuously = false };
+
+			this.Render += OnPaint;
+
+			this.Loaded += SKGLElement_Loaded;
+			this.Unloaded += SKGLElement_Unloaded;
+
+#if NETCOREAPP
+			this.RegisterToEventsDirectly = false;
+#endif
+
+			Start(settings);
+		}
+
+		private void SKGLElement_Unloaded(object sender, RoutedEventArgs e)
+		{
+			Release();
+		}
+		private void SKGLElement_Loaded(object sender, RoutedEventArgs e)
+		{
+			InvalidateVisual();
+		}
+
+		public SKSize CanvasSize => lastSize;
+
+		public GRContext GRContext => grContext;
+
+		[Category("Appearance")]
+		public event EventHandler<SKPaintGLSurfaceEventArgs> PaintSurface;
+
+		private SKSizeI GetSize()
+		{
+			var currentWidth = ActualWidth;
+			var currentHeight = ActualHeight;
+
+			if (currentWidth < 0 ||
+				currentHeight < 0)
+			{
+				currentWidth = 0;
+				currentHeight = 0;
+			}
+
+			PresentationSource source = PresentationSource.FromVisual(this);
+
+			double dpiX = 1.0;
+			double dpiY = 1.0;
+			if (source != null)
+			{
+				dpiX = source.CompositionTarget.TransformToDevice.M11;
+				dpiY = source.CompositionTarget.TransformToDevice.M22;
+			}
+
+			return new SKSizeI((int)(currentWidth * dpiX), (int)(currentHeight * dpiY));
+		}
+
+		protected override void OnRender(DrawingContext drawingContext)
+		{
+			if (grContext != null)
+			{
+				grContext.ResetContext();
```

Comment:
I am thinking about this one... GL has a "make current" concept... I wonder if the draw method is supposed to make the context current before drawing.

I see this which looks familiar, but I have no idea how to get access to this one: https://github.com/opentk/GLWpfControl/blob/516bf697e9c786ab3432156d75e1dfca67d3a18c/src/GLWpfControl/DXGLContext.cs#L127

Reply:
What I tried before was to take a variant of that code you linked from inside OpenTK (that creates the shared context) to create additional contexts, and one per SKGLElement, Then each would make current before rendering the frame. IIRC that worked. But it required essentially recreating the context creation outside of OpenTK. I can't remember if I hit another limitation with it. This would probably work, I'm guessing, but would mean we are creating one GL context per control, and I'm not sure if that is desirable? This may be what the WF is doing under the covers though due to each being a separate HWND presumably? Not sure.

Anyhow, using the shared gl context from each SKGLElement was _not_ working though without resetting the graphics context, since there were left over bits of state (maybe clipping stages?) that were crosstalking between the SKElements.

I'll see if I can find the old code that I was trying with the multiple gl contexts. I believe I posted it to discord in the past when I was asking for advice. The GL context creation was some variant of what is happening in that code you linked though.

Reply:
We can merge this with the release for now, and then in a separate PR benchmark and see if switching context is better. I don't think we need to make a new context, but somehow we need to make sure the one for the control is make current before drawing.

When you create a new GRContext, it calls make current internally, so this is what is probably making it work.

Reply:
Maybe there is a bug in the GL WPF Control, so we can also investigate that path. But, that can be done later.